76 datetime inputs

loffleraSMH opened the pull request:
Closes #76 

**Main changes**
- I added a new utility function in `utils.R` that's called `convert_dt`: It converts any date-time variables into the format expected by other Rgemini functions (typically: “ymd HM”, which is used as the default format by `convert_dt`) and returns warning messages informing users about invalid/missing date-times and (if relevant) about entries with missing timestamps/timestamps that are 00:00
- I replaced all `ymd_hm` calls with `convert_dt` (even in cases where it’s not technically needed, but just to standardize the warning messages about invalid/missing date-times)
- Removed any `lubridate` imports where they are not needed anymore
- Updated `@params` sections (where necessary) to indicate acceptable formats (character or POSIXct POSIXt)
- These changes affect the following functions: `length_of_stay()`, `er_los()`, `day_time_of_admission()`, `hospital_fiscal_year()`, `stat_holidays_ON()`, `episodes_of_care()`, `readmission()`, `daily_census()`, `icu_entry()`, `icu_los()`, `mlaps()`
- All functions should now run even when users provide non-character/pre-processed date-time inputs. I’ve tested all outputs of these functions to make sure the results are as expected/same as before.
- Note: For `n_routine_bloodwork()`, `n_imaging()`, and `n_rbc_transfusions()` I haven’t made any changes yet to avoid merge conflicts with Yishan’s PR for issue #125. I will open a separate issue for those functions because a lot of the date-times are currently handled within the SQL query (so we’d need to refactor those functions first before we can apply `convet_dt()`).

**For a few functions, I made some additional changes:**
- `er_los()`: As discussed with Yishan, I made a small change to this function to return genc_ids with an entry in er but missing/invalid ER date-time as `er_los = NA` (instead of `er_los = 0`). This only has a minimal impact on the function output (<1% of all entries in the er table are affected by this). I’ve also added a warning about this in the function and updated the documentation accordingly.
- `icu_los()`: Similar to `er_los`, any genc_ids with a missing/invalid date-time in ipscu are now returned as `icu_los = NA`.
- `icu_entry()`: I discussed this with Bijun, for any time-sensitive ICU flags (e.g., ICU within 24h), we now return genc_ids with 0 valid ICU admission date-times as NA. For any genc_ids with at least 1 valid date-time, any valid entries are considered in the calculation. I’ve updated the documentation accordingly. 
- In `day_time_of_admission()`: I switched the code to standard lubridate functions (instead of str operations) to allow for more flexible date-time formats and to make sure timestamps with 00:00 are handled appropriately (the previous version returned any entries with 00:00 as NA for daytime). Otherwise, the output of the current versions is the same as before.

Here is an example for how to test the code for `length_of_stay()`:
```
ipadmdad <- dbGetQuery(db, "select * from admdad LIMIT 100") %>% data.table() 

# pre-process date-times prior to running length_of_stay (would break old version but should work with new version)
ipadmdad[, discharge_date_time := ymd_hm(discharge_date_time)]
ipadmdad[, admission_date_time := ymd_hm(admission_date_time)]
# introduce some missing date-times (should be shown in warning message when calling new version of length_of_stay
ipadmdad[1:10, discharge_date_time := NA]

# source relevant functions
source('H:/GitHub/Rgemini/R/length_of_stay.R')
source('H:/GitHub/Rgemini/R/utils.R')

los <- length_of_stay(ipadmdad)
```

guoyi-smh commented on R/icu_los.R:
```diff
@@ -50,10 +50,16 @@
+#' Encounter IDs in the `ipscu` table that have any missing/invalid
+#' `scu_admit_date_time` or `scu_discharge_date_time` will be returned with
+#' `icu_los = NA`. Some of those entries have valid date information (but no
+#' timestamp). Users may choose to impute missing timestamps prior to running
```
Well explained. And I think it would be better to add the missing timestamp sentence to the warning in case the user does not read this documentation carefully, they can at least notice it in the warning, as this is currently a issue affecting a whole site and this site is included in most projects.

loffleraSMH replied:
Good point, I updated the warning message to `"Identified a total of XXX genc_ids with at least 1 missing/invalid 'scu_admit_date_time' or 'scu_discharge_date_time'. These entries will be returned as 'icu_los = NA'.
 Some of these entries might have valid date information (but no timestamp). Please carefully check the 'ipscu' table and consider whether it might be useful to impute missing timestamps."` Let me know if this is better.

guoyi-smh reviewed (commented):
@loffleraSMH Thank you for the great work! This new function looks great, it is well designed and it takes care of the common checks required in our daily work. I also tested all functions that were affected and they all work well.

loffleraSMH commented on R/utils.R:
```diff
@@ -652,8 +652,212 @@ mapping_message <- function(what, addtl = NULL) {
+  if (any(grepl("hm", orders, ignore.case = TRUE))) {
+    if (check_ts_zeros == TRUE) {
+      n_zeros <- sum(
+        grepl(" 00:00", as.character(dt_var)) | nchar(as.character(dt_var)) < 12
```
Good catch! I just pushed a commit to ignore missing values on this line, and I also added a unit test for this. While fixing this, I also realized my previous version didn't correctly exclude non-missing values from the denominator in checks 2-4 (so the warnings saying "... of all **non-missing** entries" returned incorrect percentages). This is resolved now too.

loffleraSMH commented on R/er_los.R:
```diff
@@ -68,10 +66,29 @@ er_los <- function(cohort, er) {
-    left_er_date_time = ymd_hm(left_er_date_time)
+    triage_date_time = convert_dt(
+      triage_date_time, addtl_msg = ""
+    ),
```
When running the function with `addtl_msg = NULL`, the generic default warning will be shown, which simply says `"Please carefully consider how to deal with missing/invalid date-time entries and perform any additional pre-processing prior to running the function (e.g., impute missing dates/timestamps etc.)"`. By specifying `addtl_msg = ""` here, I'm basically suppressing that generic warning, and instead, I'm showing a more specific warning within `er_los` that clarifies how these values are handled (see line ~80). 
I could in theory use that warning message on line ~80 as an input for `addtl_msg` but 1) it would be a bit repetitive in case the `er` input contains missing/invalid values in both `left_er_date_time` and `triage_date_time` (the message would be shown twice in that case) and 2) I wanted to mention the unique entries (i.e., `genc_ids` in this case) that are affected by missing/invalid entries in either variable. 
Does this clarify your question? Let me know if you think there is a better way of handling this.

guoyi-smh replied:
Thank you for the detail explanation! Very well-considered!

loffleraSMH commented on R/icu_entry.R:
```diff
@@ -124,25 +129,62 @@ icu_entry <- function(cohort, ipscu, as_outcome = FALSE, exclude_cutoff = 0, ent
+    warning(
+      paste(
+        "Identified a total of", nrow(all_missing),
+        "genc_ids in the ICU table that have 0 entries with valid `scu_admit_date_time`.\n",
```
Great suggestion, I found the initial version confusing too XD. I just made a small edit to your suggestion. Is this ok? 
```suggestion
        "genc_ids where all ICU entries have a missing/invalid `scu_admit_date_time`.\n",
```

loffleraSMH commented:
Thanks so much for the review and the helpful feedback @guoyi-smh! I fixed the bug and implemented your other suggestions. Let me know if there are any other changes you'd like me to make.
